Handle when an error occurs in a beforeEach block of a mocha test.

hswolff opened the pull request:
Hello, it's me again! Apologies for the PR's without issues!

This adds support to surface when an error occurs in a beforeEach block, before any individual tests run.

Without this change this is what is shown:

![image](https://user-images.githubusercontent.com/322576/34232836-38a967aa-e5b0-11e7-8f86-4ea81ba2e2f6.png)

With this change we show:

![image](https://user-images.githubusercontent.com/322576/34232842-4990dc88-e5b0-11e7-8e08-604772117d83.png)

rogeliog commented on src/runMocha.js:
```diff
@@ -16,7 +16,10 @@ const runMocha = ({ config, testPath, globalConfig }, workerCallback) => {
-      runner.on('fail', test => failures.push(test));
+      runner.on('fail', (test, err) => {
+        test.err = err;
+        failures.push(test);
```
I wonder if we should make them all `Set` instead of `Array`, and then simply call `tests.add(test)` here

rogeliog commented:
Thanks for reporting and sorry for the late response, I was sick

hswolff commented:
No worries about being sick! Appreciate the update! 

Just pushed change to using Set, let me know if that's what you had in mind otherwise I'm happy to update it to your liking.

ljharb commented on src/runMocha.js:
```diff
@@ -9,18 +9,21 @@ const runMocha = ({ config, testPath, globalConfig }, workerCallback) => {
-      const pending = [];
-      const failures = [];
-      const passes = [];
+      const tests = new Set();
```
Mocha 3 works down to node versions that lack Set entirely; if we’re going to use it, let’s be sure to be using a polyfill.

hswolff replied:
Gadzooks, mocha 3 supports all the way back to node 0.10.

My personal preference would be to either:

- Revert this last commit and keep functionality w/o Set
- Bump mocha to v4.

ljharb replied:
The former is much better.

hswolff replied:
I shall await @rogeliog's decision. (This was an async comment, cuz I _await_'ed in it. I hope my humor transcends the coldness of this comment box. :))

rogeliog replied:
I agree with @ljharb

hswolff commented:
Reverted the change to Sets for node compatibility. 

I also included another small change that hopefully is ok to include in this PR. I noticed (when trying to use https://github.com/palmerj3/jest-junit) that the type of `failureMessages` wasn't what that library was expected.

I did some looking and the [type object that is defined in Jest](https://github.com/facebook/jest/blob/3db5428ead2359206148c122f66f6badb1185b7f/types/TestResult.js#L94) also expects the property to be an Array<string>. So I fixed that here too.

rogeliog commented:
Thanks a lot!
